### PR TITLE
Introduce navigation references and link generator

### PR DESCRIPTION
**Pull request by csillag**

### CODE REVIEW

**Comment by csillag** [comment]
Wait, I got a few more ideas under the :shower: today...

**Comment by gergely-ujvari** [comment]
This PR is too technical. I need at least a basic description / introduction what is changed, introduced and what are they good for.

**Comment by gergely-ujvari** [comment]
This PR needs a description what is happening, why is this good for. This PR goes deep into technical details without any handle or help how to understand it. @csillag : I need your help here.

**Comment by gergely-ujvari** [comment]
A high-level description for this PR can be found in the comments of https://goessntl.atlassian.net/browse/VS-109

File: examples/src/MenuStructure.tsx [hunk 1]
```diff
@@ -38,6 +44,9 @@ export const mainMenu: StateSpace<UIFragment, UIFragmentSpec, {}> = [
             main: MemTableAnt,
         },
         tokenMapping: ['groupId', 'objectId'],
+        // This is where we attach this scheme to this part of the menu tree, meaning that we will
+        // be able to address these states without knowing the path to here.
+        navRef: memTableRef,
```
Comment (gergely-ujvari) [inline examples/src/MenuStructure.tsx]: A question: the navRef also implicitly defines the tokens, so you have to define them twice. Can this be streamlined more?

File: examples/src/MenuStructure.tsx [hunk 2]
```diff
@@ -46,6 +55,34 @@ export const mainMenu: StateSpace<UIFragment, UIFragmentSpec, {}> = [
...
+                // (This only works if this schema has been attached to somewhere in the menu tree.)
+                navRefCall: memTableRef.call({ groupId: 'foo', objectId: 'William8' }),
+            },
+            {
+                key: 'encodedJump',
+                label: 'Encoded jump',
+                displayOnly: true,
+                // Here, we are going to use an encoded link, which stores all in info belonging to
+                // a NavRef call, without knowing anything about the path.
+                pathTokens: ['redirect', memTableRef.call({ groupId: 'bar', objectId: 'Mary1' }).toString()],
```
Comment (gergely-ujvari) [inline examples/src/MenuStructure.tsx]: I presume that the `redirect` pathToken is a special token. Will the system warn us, if'd try to manually redefine this token?

File: packages/moxb/package.json [hunk 3]
```diff
@@ -49,12 +49,14 @@
     "tslib": "^1.9.0"
   },
   "devDependencies": {
+    "@types/atob": "2.1.2",
```
Comment (gergely-ujvari) [inline packages/moxb/package.json]: `@types/atob` is inluded here, but in the dependencies `btoa-lite` is imported. Are they using the same ts type definitions?

File: packages/moxb/src/globals.d.ts [hunk 4]
```diff
@@ -1 +1,3 @@
 declare module 'pretty';
+
+declare module 'btoa-lite';
```
Comment (gergely-ujvari) [inline packages/moxb/src/globals.d.ts]: Is this okay, just to declare an empty module? What is `@types/atob ` is used for?

File: packages/moxb/src/routing/navigation-references/LinkGeneratorImpl.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        }
+
+        // Compose the various representations of the result
+        const linkProps: CoreLinkProps = {
+            to: pathTokens,
+        };
+        const location = this._schema.getLocation(noLocation, linkProps.to!, {});
+
+        const urlProps: LocationToUrlProps = {
+            protocol: 'http',
```
Comment (gergely-ujvari) [inline packages/moxb/src/routing/navigation-references/LinkGeneratorImpl.ts]: Why are we using the example.com example urlProps here?

File: packages/moxb/src/routing/navigation-references/LinkGeneratorImpl.ts [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            pathAndQueries,
+            fullUrl,
+        };
+    }
+
+    tryGoTo(navRefCall: NavRefCall<any>, updateMethod: UpdateMethod, callback?: SuccessCallback) {
+        const { locationManager } = this.props;
+        if (!locationManager) {
+            throw new Error(
+                "This LinkGenerator hasn't been passed a LocationManager, so it can only generate links, but not go to them!"
```
Comment (gergely-ujvari) [inline packages/moxb/src/routing/navigation-references/LinkGeneratorImpl.ts]: I think "but cannot go to them" sounds better than "not go to them"

File: packages/moxb/src/routing/navigation-references/NavRefImpl.ts [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        const { linkGenerator } = this.props;
+        if (!linkGenerator) {
+            throw new Error("This NavRef has been initialized without a LinkGenerator, so we can't jump to links.");
+        }
+        linkGenerator.doGoTo(this.call(tokens), updateMethod);
+    }
+}
+
+/**
+ * Use this to define a new NavRef schema for addressing a prt of the application.
```
Comment (gergely-ujvari) [inline packages/moxb/src/routing/navigation-references/NavRefImpl.ts]: Typo (prt)

**Comment by gergely-ujvari** [comment]
I tried out the moxb example, and the encoded link doesn't seems to be working.
This is the encoded link: `http://localhost:1234/src/indexAnt.html?path=/redirect/eyJrZXkiOiJtZW1UYWJsZSIsInRva2VucyI6eyJncm91cElkIjoiYmFyIiwib2JqZWN0SWQiOiJNYXJ5MSJ9fQ%3D%3D`

Result:
```
Oops! This redirect doesn't seem to be working.
Common footer text
```

File: examples/src/MenuStructure.tsx (hunk 2 above)
Comment (csillag) [reply examples/src/MenuStructure.tsx]: No, redirect is not special.

File: packages/moxb/src/globals.d.ts (hunk 4 above)
Comment (csillag) [reply packages/moxb/src/globals.d.ts]: This basically means "any".
Unfortunately I found no ready-made typedefs, and I didn't find it to be important enough to warrant creating one.

File: packages/moxb/src/routing/navigation-references/NavRefImpl.ts (hunk 7 above)
Comment (csillag) [reply packages/moxb/src/routing/navigation-references/NavRefImpl.ts]: thx, done.

**Pull request merged**